<center>

---


---

</center>


<center>

---

# **Pandas**
---



<img src="https://drive.google.com/uc?id=1j2U9Wy535lSBdrxTCtCyseo44DWWF3gO" width="300px" height="150">
</center>

<center>

## **Introducción a pandas**

</center>

`pandas` es un paquete construido sobre `numpy`. El objeto principal de este paquete es el **DataFrame**, que son conjuntos de datos multidimensionales con etiquetas de fila y columna. Además de ofrecer una interfaz para datos etiquetados, pandas también tiene potentes funciones de manipulación de datos.


Las principales características de esta librería son:

- Define nuevas estructuras de datos basadas en los arrays de la librería NumPy pero con nuevas funcionalidades.
- Permite leer y escribir fácilmente ficheros en formato CSV, Excel y bases de datos SQL.
- Permite acceder a los datos mediante índices o nombres para filas y columnas.
- Ofrece métodos para reordenar, dividir y combinar conjuntos de datos.
- Permite trabajar con series temporales.
- Realiza todas estas operaciones de manera muy eficiente.
- Tipos de datos de Pandas
- Pandas dispone de tres estructuras de datos diferentes:

    + Series: Estructura de una dimensión.
    + DataFrame: Estructura de dos dimensiones (tablas).
    + Panel: Estructura de tres dimensiones (cubos).
Estas estructuras se construyen a partir de arrays de la librería NumPy, añadiendo nuevas funcionalidades

### **Instalando `pandas`**

La instalación de `pandas` requiere que primero tengas instalado `numpy`. Para instalarlo, utilice el comando
```python
# Instalación usando pip
!pip install pandas

# Instalación usando conda
conda install pandas
```

In [ ]:
# Si no tiene instalado pandas
#!pip install pandas

Después de la instalación, se puede importar. La versión se puede comprobar de la siguiente manera:

In [ ]:
# Importar librerias
import pandas as pd
import numpy as np

In [ ]:
# Pandas Version
pd.__version__

'1.1.5'

## **Objetos de Pandas**

<img src ="https://drive.google.com/uc?id=10wU_f9PK5_nxb-mpPoT6TZLw_AO5IfxJ" width="600px" height="300px" >

Vamos a trabajar con el famoso [Titanic dataset](https://www.kaggle.com/c/titanic/data). Estos son los campos disponibles:

---
- `passengerId`: Un ID numérico asignado a cada pasajero
- `survived`: Si el pasajero sobrevivió (1), o no (0)
- `pclass`: La clase en la que estaba el pasajero
- `nombre`: el nombre del pasajero
- `Sex`: El género del pasajero - hombre o mujer
- `Age`: La edad del pasajero (incluye valores fraccionarios)
- `Sibsp`: El número de hermanos y cónyuges que el pasajero tenía a bordo
- `Parch`: El número de padres e hijos que el pasajero tenía a bordo
- `Ticket`: El número de billete del pasajero
- `Tarifa`: Cuánto pagó el pasajero por el billete
- `Cabin`: La cabina en la que estaba el pasajero
- `Embark`: Donde el pasajero embarcó en el Titanic

---

Cargar datos
Originalmente un enlace copiado desde drive viene así:

https://drive.google.com/file/d/1GxTVun04y5gmzG1VtXoIdum7-PN1Uy23/view?usp=sharing


Nosotros debemos copiarlo así:

https://drive.google.com/uc?id=1GxTVun04y5gmzG1VtXoIdum7-PN1Uy23



In [ ]:
url='https://drive.google.com/uc?id=1MhFgQkQinmH3EoPeZ7L1sOyfJNG4j5mR'
titanic=pd.read_csv(url)
titanic

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0              1         0       3  ...   7.2500   NaN         S
1              2         1       1  ...  71.2833   C85         C
2              3         1       3  ...   7.9250   NaN         S
3              4         1       1  ...  53.1000  C123         S
4              5         0       3  ...   8.0500   NaN         S
..           ...       ...     ...  ...      ...   ...       ...
886          887         0       2  ...  13.0000   NaN         S
887          888         1       1  ...  30.0000   B42         S
888          889         0       3  ...  23.4500   NaN         S
889          890         1       1  ...  30.0000  C148         C
890          891         0       3  ...   7.7500   NaN         Q

[891 rows x 12 columns]

In [ ]:
titanic.index

RangeIndex(start=0, stop=891, step=1)

## Partes de un DataFrame
Para entender mejor los objetos DataFrame, es útil saber que constan de tres componentes, almacenados como atributos:

- `.values`: Un array bidimensional de valores NumPy.
- `.columns` : Un índice de columnas: los nombres de las columnas.
- `.index:` Un índice de las filas: los números de las filas o los nombres de las filas.

Normalmente se puede pensar en los índices como una lista de cadenas o números, aunque el tipo de datos `Index` de pandas permite opciones más sofisticadas. 

- **Nombre de las columnas**

In [ ]:
titanic.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

- **Muestra los valores**  

In [ ]:
titanic.values

array([[1, 0, 3, ..., 7.25, nan, 'S'],
       [2, 1, 1, ..., 71.2833, 'C85', 'C'],
       [3, 1, 3, ..., 7.925, nan, 'S'],
       ...,
       [889, 0, 3, ..., 23.45, nan, 'S'],
       [890, 1, 1, ..., 30.0, 'C148', 'C'],
       [891, 0, 3, ..., 7.75, nan, 'Q']], dtype=object)

In [ ]:
titanic.index

RangeIndex(start=0, stop=891, step=1)

## **Ordenar las filas**
Encontrar datos interesantes en un DataFrame suele ser más fácil si se cambia el orden de las filas. Puede ordenar las filas pasando un nombre de columna a `.sort_values()`

En los casos en los que las filas tienen el mismo valor (esto es común si se ordena en una variable categórica), es posible que desee romper los empates ordenando en otra columna. Puede ordenar en múltiples columnas de esta manera pasando una lista de nombres de columnas.


| Ordenar en ... | Sintaxis |
|---|---|
|una columna| `df.sort_values("raza")`|
|varias columnas |`df.sort_values(["raza", "peso_kg"])`|

Combinando `.sort_values()` con `.head()`, se puede responder a preguntas del tipo "¿Cuáles son los principales casos en los que...?".


In [ ]:
titanic.sort_values("Age")

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
803          804         1       3  ...   8.5167   NaN         C
755          756         1       2  ...  14.5000   NaN         S
644          645         1       3  ...  19.2583   NaN         C
469          470         1       3  ...  19.2583   NaN         C
78            79         1       2  ...  29.0000   NaN         S
..           ...       ...     ...  ...      ...   ...       ...
859          860         0       3  ...   7.2292   NaN         C
863          864         0       3  ...  69.5500   NaN         S
868          869         0       3  ...   9.5000   NaN         S
878          879         0       3  ...   7.8958   NaN         S
888          889         0       3  ...  23.4500   NaN         S

[891 rows x 12 columns]

In [ ]:
titanic.sort_values('Age',ascending=False)

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
630          631         1       1  ...  30.0000   A23         S
851          852         0       3  ...   7.7750   NaN         S
493          494         0       1  ...  49.5042   NaN         C
96            97         0       1  ...  34.6542    A5         C
116          117         0       3  ...   7.7500   NaN         Q
..           ...       ...     ...  ...      ...   ...       ...
859          860         0       3  ...   7.2292   NaN         C
863          864         0       3  ...  69.5500   NaN         S
868          869         0       3  ...   9.5000   NaN         S
878          879         0       3  ...   7.8958   NaN         S
888          889         0       3  ...  23.4500   NaN         S

[891 rows x 12 columns]

- Ordene  `titanic` primero por edad (ascendente) y luego por clase (descendente).

In [ ]:
titanic.sort_values(['Age', 'Pclass' ], ascending=[True, False])

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
803          804         1       3  ...   8.5167   NaN         C
755          756         1       2  ...  14.5000   NaN         S
469          470         1       3  ...  19.2583   NaN         C
644          645         1       3  ...  19.2583   NaN         C
78            79         1       2  ...  29.0000   NaN         S
..           ...       ...     ...  ...      ...   ...       ...
766          767         0       1  ...  39.6000   NaN         C
793          794         0       1  ...  30.6958   NaN         C
815          816         0       1  ...   0.0000  B102         S
839          840         1       1  ...  29.7000   C47         C
849          850         1       1  ...  89.1042   C92         C

[891 rows x 12 columns]

In [ ]:
titanic.shape

(891, 12)

## **Subconjunto de columnas**
Cuando se trabaja con datos, es posible que no se necesiten todas las variables del conjunto de datos. Los corchetes (`[ ]`) pueden utilizarse para seleccionar sólo las columnas que le interesan en un orden que tenga sentido para usted. Para seleccionar sólo `"col_a"` del DataFrame `df`, utilice

- `df["col_a"]`

Para seleccionar `"col_a"` y `"col_b"` de `df`, utilice

- `df[["col_a", "col_b"]]`

- Crea un DataFrame llamado  que contenga sólo la columna `Age` de `titanic`.
- Imprima la cabeza del resultado.

In [ ]:
titanic['Age']

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888     NaN
889    26.0
890    32.0
Name: Age, Length: 891, dtype: float64

In [ ]:
# Escriba su código
titanic['Sex']


0        male
1      female
2      female
3      female
4        male
        ...  
886      male
887    female
888    female
889      male
890      male
Name: Sex, Length: 891, dtype: object

- Cree un DataFrame llamado  que contenga sólo las columnas `Sex` , `Survived`y `Pclass` de `titanic`, en ese orden.
- Imprima la cabeza del resultado.

In [ ]:
# escriba su código


Sex  Survived  Pclass
0    male         0       3
1  female         1       1
2  female         1       3
3  female         1       1
4    male         0       3
5    male         0       3
6    male         0       1
7    male         0       3
8  female         1       3
9  female         1       2

## Subconjunto de filas
Una gran parte de  en la manipulación de  datos consiste en encontrar qué partes del conjunto de datos son interesantes. Una de las técnicas más sencillas para ello consiste en encontrar un subconjunto de filas que coincidan con algún criterio. Esto se conoce a veces como **filtrado de filas** o **selección de filas**.

Hay muchas maneras de hacer un subconjunto de un DataFrame, quizás la más común es usar operadores relacionales para devolver `True` o `False` para cada fila, y luego pasar eso dentro de corchetes.


```python
perros[perros["altura_cm"] > 60]
perros[perros["color"] == "bronceado"]
```
Puede filtrar varias condiciones a la vez utilizando el operador "bitwise and", `&`.

```python
perros[(perros["altura_cm"] > 60) & (perros["color"] == "bronceado")]
```

In [ ]:
titanic[titanic["Pclass"]==1]

PassengerId  Survived  Pclass  ...     Fare        Cabin  Embarked
1              2         1       1  ...  71.2833          C85         C
3              4         1       1  ...  53.1000         C123         S
6              7         0       1  ...  51.8625          E46         S
11            12         1       1  ...  26.5500         C103         S
23            24         1       1  ...  35.5000           A6         S
..           ...       ...     ...  ...      ...          ...       ...
871          872         1       1  ...  52.5542          D35         S
872          873         0       1  ...   5.0000  B51 B53 B55         S
879          880         1       1  ...  83.1583          C50         C
887          888         1       1  ...  30.0000          B42         S
889          890         1       1  ...  30.0000         C148         C

[216 rows x 12 columns]

- Filtre pasajeros del titanic mayores  de 60 años

In [ ]:
titanic[titanic['Age']>60]

PassengerId  Survived  Pclass  ...      Fare        Cabin  Embarked
33            34         0       2  ...   10.5000          NaN         S
54            55         0       1  ...   61.9792          B30         C
96            97         0       1  ...   34.6542           A5         C
116          117         0       3  ...    7.7500          NaN         Q
170          171         0       1  ...   33.5000          B19         S
252          253         0       1  ...   26.5500          C87         S
275          276         1       1  ...   77.9583           D7         S
280          281         0       3  ...    7.7500          NaN         Q
326          327         0       3  ...    6.2375          NaN         S
438          439         0       1  ...  263.0000  C23 C25 C27         S
456          457         0       1  ...   26.5500          E38         S
483          484         1       3  ...    9.5875          NaN         S
493          494         0       1  ...   49.5042          NaN         C
545          546         0       1  ...   26.0000          NaN         S
555          556         0       1  ...   26.5500          NaN         S
570          571         1       2  ...   10.5000          NaN         S
625          626         0       1  ...   32.3208          D50         S
630          631         1       1  ...   30.0000          A23         S
672          673         0       2  ...   10.5000          NaN         S
745          746         0       1  ...   71.0000          B22         S
829          830         1       1  ...   80.0000          B28       NaN
851          852         0       3  ...    7.7750          NaN         S

[22 rows x 12 columns]

In [ ]:
a=titanic[titanic['Age']>60]
a

PassengerId  Survived  Pclass  ...      Fare        Cabin  Embarked
33            34         0       2  ...   10.5000          NaN         S
54            55         0       1  ...   61.9792          B30         C
96            97         0       1  ...   34.6542           A5         C
116          117         0       3  ...    7.7500          NaN         Q
170          171         0       1  ...   33.5000          B19         S
252          253         0       1  ...   26.5500          C87         S
275          276         1       1  ...   77.9583           D7         S
280          281         0       3  ...    7.7500          NaN         Q
326          327         0       3  ...    6.2375          NaN         S
438          439         0       1  ...  263.0000  C23 C25 C27         S
456          457         0       1  ...   26.5500          E38         S
483          484         1       3  ...    9.5875          NaN         S
493          494         0       1  ...   49.5042          NaN         C
545          546         0       1  ...   26.0000          NaN         S
555          556         0       1  ...   26.5500          NaN         S
570          571         1       2  ...   10.5000          NaN         S
625          626         0       1  ...   32.3208          D50         S
630          631         1       1  ...   30.0000          A23         S
672          673         0       2  ...   10.5000          NaN         S
745          746         0       1  ...   71.0000          B22         S
829          830         1       1  ...   80.0000          B28       NaN
851          852         0       3  ...    7.7750          NaN         S

[22 rows x 12 columns]

- filtrar Pasajeros del titanic mayores de 60 años que viajaban en primera clase

In [ ]:
titanic[(titanic['Age']>60) & (titanic['Survived']==0)]

PassengerId  Survived  Pclass  ...      Fare        Cabin  Embarked
33            34         0       2  ...   10.5000          NaN         S
54            55         0       1  ...   61.9792          B30         C
96            97         0       1  ...   34.6542           A5         C
116          117         0       3  ...    7.7500          NaN         Q
170          171         0       1  ...   33.5000          B19         S
252          253         0       1  ...   26.5500          C87         S
280          281         0       3  ...    7.7500          NaN         Q
326          327         0       3  ...    6.2375          NaN         S
438          439         0       1  ...  263.0000  C23 C25 C27         S
456          457         0       1  ...   26.5500          E38         S
493          494         0       1  ...   49.5042          NaN         C
545          546         0       1  ...   26.0000          NaN         S
555          556         0       1  ...   26.5500          NaN         S
625          626         0       1  ...   32.3208          D50         S
672          673         0       2  ...   10.5000          NaN         S
745          746         0       1  ...   71.0000          B22         S
851          852         0       3  ...    7.7750          NaN         S

[17 rows x 12 columns]

**¿Cuantas mujeres entre 30 y 40 años viajaban en el titanic**


### **Indexadores: `loc` y `iloc`**

Pandas proporciona algunos atributos especiales *indexadores* que exponen explícitamente ciertos esquemas de indexación. No se trata de métodos funcionales, sino de atributos que exponen una forma particular de recuperar segmentos de los datos.

- El atributo `loc` permite una indexación y segmentación que siempre hace referencia al índice explícito. 

- El atributo `iloc` permite la indexación y segmentación que siempre se refiere al índice implícito de estilo Python. 

Para ilustrar la diferencia, veamos los siguientes ejemplos. 

- `loc` incluye ambos extremos de los rangos especificados, y que las columnas se especifican por su nombre. 
- `iloc` es exclusivo del extremo superior de los rangos especificados, y las columnas se especifican por índice numérico.


In [ ]:
titanic.loc[10:15,'PassengerId':'Age']

PassengerId  Survived  ...     Sex   Age
10           11         1  ...  female   4.0
11           12         1  ...  female  58.0
12           13         0  ...    male  20.0
13           14         0  ...    male  39.0
14           15         0  ...  female  14.0
15           16         1  ...  female  55.0

[6 rows x 6 columns]

In [ ]:
titanic.iloc[10:15,0:5]

PassengerId  Survived  Pclass                                  Name     Sex
10           11         1       3       Sandstrom, Miss. Marguerite Rut  female
11           12         1       1              Bonnell, Miss. Elizabeth  female
12           13         0       3        Saundercock, Mr. William Henry    male
13           14         0       3           Andersson, Mr. Anders Johan    male
14           15         0       3  Vestrom, Miss. Hulda Amanda Adolfina  female

### Selección de datos

Ahora que sabemos un poco sobre cómo `pandas` indexa los datos, podemos ver cómo extraerlos. `pandas` tiene varias formas de seleccionar subconjuntos de un conjunto de datos:

In [ ]:
# slicing  por indexación implicita
titanic[10:15]

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
10           11         1       3  ...  16.7000    G6         S
11           12         1       1  ...  26.5500  C103         S
12           13         0       3  ...   8.0500   NaN         S
13           14         0       3  ...  31.2750   NaN         S
14           15         0       3  ...   7.8542   NaN         S

[5 rows x 12 columns]

In [ ]:
titanic[['PassengerId','Age','Sex']][10:15]
#titanic.shape

PassengerId   Age     Sex
10           11   4.0  female
11           12  58.0  female
12           13  20.0    male
13           14  39.0    male
14           15  14.0  female

In [ ]:
titanic[50:55]

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
50           51         0       3  ...  39.6875   NaN         S
51           52         0       3  ...   7.8000   NaN         S
52           53         1       1  ...  76.7292   D33         C
53           54         1       2  ...  26.0000   NaN         S
54           55         0       1  ...  61.9792   B30         C

[5 rows x 12 columns]

#### Filtro

In [ ]:
titanic[(titanic['Age'] > 18) & (titanic['Age'] < 50) & (titanic['Sex']=='female') ][10:20]

PassengerId  Survived  Pclass  ...      Fare        Cabin  Embarked
53            54         1       2  ...   26.0000          NaN         S
56            57         1       2  ...   10.5000          NaN         S
61            62         1       1  ...   80.0000          B28       NaN
66            67         1       2  ...   10.5000          F33         S
79            80         1       3  ...   12.4750          NaN         S
85            86         1       3  ...   15.8500          NaN         S
88            89         1       1  ...  263.0000  C23 C25 C27         S
98            99         1       2  ...   23.0000          NaN         S
100          101         0       3  ...    7.8958          NaN         S
106          107         1       3  ...    7.6500          NaN         S

[10 rows x 12 columns]

In [ ]:
titanic['Age'][10:15]

10     4.0
11    58.0
12    20.0
13    39.0
14    14.0
Name: Age, dtype: float64

In [ ]:
titanic[(titanic["Pclass"]==1) | (titanic["Survived"]==0)]
titanic.Pclass

0      3
1      1
2      3
3      1
4      3
      ..
886    2
887    1
888    3
889    1
890    3
Name: Pclass, Length: 891, dtype: int64

## Anatomia de un `DataFrame`

<center>
<img src="https://drive.google.com/uc?id=1mU9TqPnxkkP15nX225mMjSg8kvcVdFMA" width="400px" height="300px">
</center>


### `concat()`

`pandas` tiene una función, `pd.concat()`, que tiene una sintaxis similar a ``np.concatenate`` pero contiene diferentes argumentos. Se puede utilizar para una simple concatenación de ``Series`` o ``DataFrames``:

In [ ]:
df1 = titanic[(titanic.Age > 18) & (titanic['Age'] < 50)][10:15][['PassengerId','Survived']]
df1.reset_index(inplace = True)
df1

index  PassengerId  Survived
0     21           22         1
1     23           24         1
2     25           26         1
3     27           28         0
4     30           31         0

In [ ]:
df2 = titanic.iloc[10:15,2:5]
df2.reset_index(inplace = True)
df2

index  Pclass                                  Name     Sex
0     10       3       Sandstrom, Miss. Marguerite Rut  female
1     11       1              Bonnell, Miss. Elizabeth  female
2     12       3        Saundercock, Mr. William Henry    male
3     13       3           Andersson, Mr. Anders Johan    male
4     14       3  Vestrom, Miss. Hulda Amanda Adolfina  female

In [ ]:
pd.concat([df1, df2],sort=True, axis=1)

index  PassengerId  ...                                  Name     Sex
0     21           22  ...       Sandstrom, Miss. Marguerite Rut  female
1     23           24  ...              Bonnell, Miss. Elizabeth  female
2     25           26  ...        Saundercock, Mr. William Henry    male
3     27           28  ...           Andersson, Mr. Anders Johan    male
4     30           31  ...  Vestrom, Miss. Hulda Amanda Adolfina  female

[5 rows x 7 columns]

In [ ]:
pd.concat([df1, df2],ignore_index=True,sort=True, axis=0)

Name  PassengerId  ...  Survived index
0                                   NaN         22.0  ...       1.0    21
1                                   NaN         24.0  ...       1.0    23
2                                   NaN         26.0  ...       1.0    25
3                                   NaN         28.0  ...       0.0    27
4                                   NaN         31.0  ...       0.0    30
5       Sandstrom, Miss. Marguerite Rut          NaN  ...       NaN    10
6              Bonnell, Miss. Elizabeth          NaN  ...       NaN    11
7        Saundercock, Mr. William Henry          NaN  ...       NaN    12
8           Andersson, Mr. Anders Johan          NaN  ...       NaN    13
9  Vestrom, Miss. Hulda Amanda Adolfina          NaN  ...       NaN    14

[10 rows x 6 columns]

## Agregación y agrupación

Una parte esencial del análisis de muchos datos es la creación de resúmenes eficientes de los mismos. Las agregaciones computacionales, como ``sum()``, ``mean()``, ``median()``, ``min()`` y ``max()``, nos permiten utilizar relativamente pocos números para darnos una idea de la naturaleza de un conjunto de datos potencialmente grande.

En esta sección, exploraremos las agregaciones en `pandas`, desde operaciones simples similares a las que hemos visto en las matrices `numpy`, hasta operaciones más sofisticadas basadas en el concepto de `groupby()`.


In [ ]:
titanic['Age'].mean()

29.69911764705882

In [ ]:
titanic.groupby('Sex').size()

Sex
female    314
male      577
dtype: int64

In [ ]:
titanic.groupby('Sex').agg({'Age':['min','mean','max']})

Age                 
         min       mean   max
Sex                          
female  0.75  27.915709  63.0
male    0.42  30.726645  80.0

In [ ]:
titanic.describe(include=object)

Name   Sex Ticket        Cabin Embarked
count                            891   891    891          204      889
unique                           891     2    681          147        3
top     Gustafsson, Mr. Johan Birger  male   1601  C23 C25 C27        S
freq                               1   577      7            4      644

### `crosstab()`

Una de las funciones útiles de `pandas` es `crosstab()`. Esta función nos permite separar nuestros datos en cubos utilizando las clases de algunas variables categóricas, y luego construir una tabla que muestra el número de puntos de datos en cada cubo:

In [ ]:
pd.crosstab(titanic['Sex'], titanic['Pclass'])

Pclass    1    2    3
Sex                  
female   94   76  144
male    122  108  347

### Tablas dinámicas

Son como las tablas dinámicas de Excel. Aquí la utilizamos para ver la edad media de los pasajeros, segregada por clase y género:

In [ ]:
titanic.pivot_table(values='Age',index='Sex',columns='Pclass', aggfunc=np.mean)

Pclass          1          2          3
Sex                                    
female  34.611765  28.722973  21.750000
male    41.281386  30.740707  26.507589

## Explorando los datos

A medida que avanzamos en el conjunto de datos, conocer algunas funciones que nos proporcionan las `pandas` para facilitarnos la vida es crucial. Aquí queremos identificar los tipos de datos de nuestro conjunto de datos, si tenemos nulos en nuestro conjunto de datos, etc. Algunas funciones que nos pueden ayudar para esto son

```python
df.head()
df.tail()
df.info()
df.shape
df.columns
df.describe()
df.value_counts()
df.unique()
df.nunique()
```

- ver las primeras dos filas


In [ ]:
titanic.head(2)

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C

[2 rows x 12 columns]

- ver las últimas dos filas

In [ ]:
titanic.tail(2)

PassengerId  Survived  Pclass  ...   Fare Cabin  Embarked
889          890         1       1  ...  30.00  C148         C
890          891         0       3  ...   7.75   NaN         Q

[2 rows x 12 columns]

- Cuantas filas y columnas hay en el dataset


In [ ]:
titanic.shape

(891, 12)

- Tipos de datos y cantidad de valores nulos en el conjunto de datos


In [ ]:
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


- Otra forma de ver los valores nulos por columna


In [ ]:
titanic.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

- Otra forma de ver el tipo de datos

In [ ]:
titanic.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [ ]:
# Información sobre las columnas numéricas
titanic.describe()

PassengerId    Survived      Pclass  ...       SibSp       Parch        Fare
count   891.000000  891.000000  891.000000  ...  891.000000  891.000000  891.000000
mean    446.000000    0.383838    2.308642  ...    0.523008    0.381594   32.204208
std     257.353842    0.486592    0.836071  ...    1.102743    0.806057   49.693429
min       1.000000    0.000000    1.000000  ...    0.000000    0.000000    0.000000
25%     223.500000    0.000000    2.000000  ...    0.000000    0.000000    7.910400
50%     446.000000    0.000000    3.000000  ...    0.000000    0.000000   14.454200
75%     668.500000    1.000000    3.000000  ...    1.000000    0.000000   31.000000
max     891.000000    1.000000    3.000000  ...    8.000000    6.000000  512.329200

[8 rows x 7 columns]

In [ ]:
#Información sobre las columnas categoricas 
titanic.describe(include = object)

Name   Sex Ticket        Cabin Embarked
count                            891   891    891          204      889
unique                           891     2    681          147        3
top     Gustafsson, Mr. Johan Birger  male   1601  C23 C25 C27        S
freq                               1   577      7            4      644

- Valores únicos por columna excluyendo los nulos


In [ ]:
titanic.nunique()

PassengerId    891
Survived         2
Pclass           3
Name           891
Sex              2
Age             88
SibSp            7
Parch            7
Ticket         681
Fare           248
Cabin          147
Embarked         3
dtype: int64

In [ ]:
# Valores únicos por columna, incluidos los nulos
titanic.nunique(dropna=False)

PassengerId    891
Survived         2
Pclass           3
Name           891
Sex              2
Age             89
SibSp            7
Parch            7
Ticket         681
Fare           248
Cabin          148
Embarked         4
dtype: int64

In [ ]:
# Counting
titanic.count()

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64

In [ ]:
# Valor más frecuente de la serie
titanic['Embarked'].mode()

# Valores únicos de la columna "Embarcado''
#titanic['Embarked'].unique()

0    S
dtype: object

In [1]:
# Cuántos registros tenemos por categoría de la columna 'Embarcado' sin valores nulos


In [ ]:
# Cuántos registros tenemos por categoría de la columna 'Embarcado' con valores nulos
titanic['Embarked'].value_counts(dropna=False)

S      644
C      168
Q       77
NaN      2
Name: Embarked, dtype: int64

In [ ]:
# Nombre de las columnas en el dataset
titanic.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [ ]:
titanic.columns = ['Passenger_Id', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked']
titanic = titanic.rename(columns={'Passenger_Id':'PassengerId', "Sex":"Sexo", "Ticket":"Tiquete"})
titanic.head()

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S
3            4         1       1  ...  53.1000  C123         S
4            5         0       3  ...   8.0500   NaN         S

[5 rows x 12 columns]

In [ ]:
titanic[(titanic['Embarked']=='S') & (titanic['Sexo']=='female')][['Name','Age']].head(3)

Name   Age
2                             Heikkinen, Miss. Laina  26.0
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  27.0

### **Ejercicio 1:**

Utilizando las columnas `"sexo"`, `"edad"` y `"tarifa"`, responde a las siguientes preguntas:

1. ¿Cuántos hombres y mujeres tenemos en nuestro conjunto de datos?
2. ¿Cuál es la edad más frecuente en nuestro conjunto de datos?
3. ¿Cuál es la edad más frecuente en las mujeres?
4. ¿Cuál es la edad más frecuente para los hombres?
5. ¿Cuál es el precio más común del billete?
6. ¿Es el precio del billete más común el mismo para hombres y mujeres?


- Punto 1

In [2]:
# Su código


PassengerId  Survived  Pclass  Name  ...  Tiquete  Fare  Cabin  Embarked
Sexo                                         ...                                
female          314       314     314   314  ...      314   314     97       312
male            577       577     577   577  ...      577   577    107       577

[2 rows x 11 columns]

- Punto 2

In [ ]:
# Escriba su código



0    24.0
dtype: float64

- Punto 3

In [ ]:
titanic.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sexo', 'Age', 'SibSp',
       'Parch', 'Tiquete', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [3]:
# Escriba su código


In [ ]:
titanic[(titanic["Sexo"]=="female")& (titanic["Age"]==24)]

PassengerId  Survived  Pclass  ...      Fare        Cabin  Embarked
142          143         1       3  ...   15.8500          NaN         S
199          200         0       2  ...   13.0000          NaN         S
247          248         1       2  ...   14.5000          NaN         S
293          294         0       3  ...    8.8500          NaN         S
310          311         1       1  ...   83.1583          C54         C
316          317         1       2  ...   26.0000          NaN         S
341          342         1       1  ...  263.0000  C23 C25 C27         S
345          346         1       2  ...   13.0000          F33         S
369          370         1       1  ...   69.3000          B35         C
394          395         1       3  ...   16.7000           G6         S
437          438         1       2  ...   18.7500          NaN         S
600          601         1       2  ...   27.0000          NaN         S
615          616         1       2  ...   65.0000          NaN         S
641          642         1       1  ...   69.3000          B35         C
710          711         1       1  ...   49.5042          C90         C
858          859         1       3  ...   19.2583          NaN         C

[16 rows x 12 columns]

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


- Punto 4

In [ ]:
# Escriba su código

- Punto 5

In [ ]:
# Escriba su código

- Punto 6

In [ ]:
# Escriba su código

## **Pre-procesamiento de nuestros datos**

Ahora que tenemos una idea de en qué consiste nuestro conjunto de datos, vamos a transformarlo para poder visualizar la fase. Los tipos de actividades que podemos realizar durante el **preprocesamiento** incluyen

1. **Eliminar columnas**
2. **Enriquecer (o transformar)** un conjunto de datos, añadiendo nuevas columnas calculadas en los índices
3. **Filtrar** un subconjunto de filas o columnas de un conjunto de datos según algún criterio
4. **Indexar** un conjunto de datos
5. **Agregar** datos
6. **Ordenar** las filas de un conjunto de datos según algún criterio
7. **Fusionar** los datos
8. **Pivotar** de forma que los datos que originalmente estaban dispuestos en vertical se dispongan en horizontal (aumentando el número de columnas) o viceversa (aumentando el número de filas)

entre otros.



In [ ]:
titanic.columns

- Eliminar las columnas `PassengerId`, `Name` , Cabin, no vamos a eliminar la columna del billete todavía.

In [ ]:
titanic.drop(['PassengerId','Name','Cabin'],axis=1, inplace=True)
titanic.shape

In [ ]:
# Validación nula de nuevo
titanic.isnull().sum()

In [ ]:
# En los primeros 10 registros tenemos un campo de edad nulo, lo vamos a tener como referencia para lo que vamos a hacer
titanic.head(10)

### **Valores nulos**

Dependiendo del problema, el tratamiento de los nulos puede ser manejado de diferentes maneras:

* Mantenerlos
* Borrarlos
* Dejar un valor fijo que los diferencie
* Sustituirlos por un estadístico de resumen (media, mediana, moda)
* Rellenar hacia delante o hacia atrás
* Interpolar



In [ ]:
# Delete
titanic.dropna(inplace=True)
titanic.shape

In [ ]:
# Validación nulos
titanic.isnull().sum()

In [ ]:
titanic.info()

In [ ]:
titanic.reset_index(drop=True,inplace=True)
titanic.head(10)

In [ ]:
# Deja un valor fijo que los diferencie
titanic = pd.read_csv("https://drive.google.com/uc?id=1_uYmhihRAPXkYky1G_cJri5IV0-szgjP")
titanic.drop(['PassengerId','Name','Cabin'],axis=1, inplace=True)

In [ ]:
titanic.shape

In [ ]:
titanic.fillna(-1).head(10)


In [ ]:
titanic.fillna(titanic['Age'].mean()).head(10)

In [ ]:
titanic['Age'].fillna(titanic['Age'].median()).head(10)

In [ ]:
titanic['Age'].fillna(titanic['Age'].mode()[0]).head(10)

In [ ]:
titanic['Age'].fillna(method='backfill').head(10)

In [ ]:
titanic['Age'].fillna(method='ffill').head(10)

In [ ]:
titanic.interpolate(method='polynomial', order=2).head(10)

Para nuestro ejercicio tomaremos la decisión de eliminar los nulos.

---



In [ ]:
# Drop nulls
titanic.dropna(inplace=True)
titanic.shape


In [ ]:
titanic.reset_index(drop=True , inplace=True)
titanic.head(7)

### Cambio de nombre de las columnas

Para ser prácticos, normalmente modificamos los nombres de las columnas para eliminar los espacios en blanco y los caracteres especiales, y para convertir las mayúsculas en minúsculas:

-  Cambiar el nombre de las columnas a minúsculas


In [ ]:
titanic.rename(columns = lambda col: col.lower(),inplace=True)
titanic

**Sintaxis de una función Lambda**

`lambda argumentos: expresión`

Las funciones Lambda pueden tener cualquier número de argumentos, pero solo una expresión.

In [ ]:
# Validando dataset
titanic.info()

In [ ]:
titanic.shape

### Indexación

In [ ]:
# Después de eliminar los nulos, hay que reiniciar los índices, como puedes ver siguen apareciendo de 0 a 890 
# aunque tengamos 712 filas.
titanic.reset_index(drop=True,inplace=True)
titanic

### **Agrupación**

<img src="https://drive.google.com/uc?id=1h4sbR-AXgzcdZ2eCtBpQiDGPfcF8Q-gL" width="400px" heigt="400px">

In [ ]:
titanic.groupby(['sex','survived']).size()

In [ ]:
titanic.groupby(['sex','survived']).agg({'fare':'count','age':'mean'}).reset_index()

In [ ]:
# Validar los tickets duplicados
df_tickets = titanic.groupby('ticket').size().reset_index().rename(columns={0:'count'})
df_tickets
df_tickets.sort_values(by='count',ascending=False)

In [ ]:
titanic[titanic['ticket']=='347082']

In [ ]:
titanic[titanic['ticket']=='110413']

### Ordering

In [ ]:
# Ordenar el número de registros por entradas
df_tickets.sort_values(by=["count"],ascending=False,inplace=True)
df_tickets.head(10)

In [ ]:
# Vamos a validar un ticket
titanic[titanic['ticket']=='347082']

### Merging datasets

In [ ]:
# Primero añadiremos la columna 'count' de df_tickets a nuestro conjunto de datos titanic
titanic = titanic.merge(df_tickets,left_on='ticket',right_on='ticket',how='left')
titanic.shape

In [ ]:
titanic.head(15)

In [ ]:
df_A = pd.DataFrame({
    'Fecha': ['24/06/2021', '25/06/2021', '26/06/2021'],
    'Datos_A': [1, 5, 9]
})


In [ ]:
df_A

In [ ]:
df_B = pd.DataFrame({
    'Fecha': ['24/06/2021', '25/06/2021', '26/06/2021', '27/06/2021'],
    'Datos_B': [7, 5, 3, 4]
})

In [ ]:
df_B

In [ ]:
df_A.Fecha = pd.to_datetime(df_A.Fecha)
df_B.Fecha = pd.to_datetime(df_B.Fecha)

In [ ]:
df = pd.merge(df_A, df_B, on='Fecha', how='outer')
df
#df.sort_values(by='Fecha', inplace=True)
#df[['Fecha', 'Datos_A', 'Datos_B']]

## **Melt**

La función `pd.melt()` remodela o transforma un DataFrame existente. Cambia la orientación del DataFrame de un formato ancho a uno largo.

<img src="https://drive.google.com/uc?id=19gZqC45_VptAmtGp9zZLSri2rhq85vDm" width="500px" height=" 300px" >

In [ ]:
dataframe=pd.DataFrame({'Asistencia': {0: 60, 1: 100, 2: 80,3: 78,4: 95},
                    'Nombre': {0: 'Andres', 1: 'John', 2: 'Laura',3: 'Wilson',4: 'Jorge'},
                    'porcentaje obtenido': {0: '90%', 1: '75%', 2: '82%',3: '64%',4: '45%'}})
dataframe

In [ ]:
dataframe1 = pd.melt(dataframe)
dataframe1


In [ ]:
dataframe2 = pd.melt(dataframe,  id_vars=['Nombre'])
dataframe2

In [ ]:
dataframe3 = pd.melt(dataframe, 
                     id_vars=['Nombre'], 
                     var_name = 'Performance', 
                     value_name = 'Success')
dataframe3

In [ ]:
dataframe4 = pd.melt(dataframe, id_vars=['Nombre'], value_vars='Attendance', var_name = 'Performance', value_name = 'Success')
dataframe4

In [ ]:
print(dataframe3.pivot(index='Nombre', columns='Performance', values='Success'))


R en python


In [ ]:
%load_ext rpy2.ipython

In [ ]:
%%R

x=c(4,5,7)
mean(x)
var(x)
y=c(3,6,8)
x+y


### Enriquecer el conjunto de datos

Vamos a **enriquecer** este conjunto de datos creando una nueva columna `familia`, que será "1" si el viaje se realizó con otras personas, y "0" si viajaron solos:

In [ ]:
titanic['family'] = [1 if count>1 else 0 for count in titanic['count']]

In [ ]:
titanic.head()

### Cleaning text

In [ ]:
# vamos a crear una función que nos ayude a limpiar el texto, en este caso el campo tickets
def clean_text(string):
    import re
    regex = '[^A-Za-z0-9]+'
    return re.sub(regex, '', string)

In [ ]:
titanic['ticket'] = titanic['ticket'].apply(clean_text)
titanic.head()

In [ ]:
# Para continuar borrar la columna ticket
titanic.drop('ticket',axis=1,inplace=True)
titanic.head()

### Codificación de variables categóricas

A veces, es útil crear una columna para cada categoría posible de una de nuestras variables categóricas. Para ello, utilizamos un método `pandas` llamado `get_dummies()`:

In [ ]:
titanic['sexDummies'] = titanic['sex']
titanic = pd.get_dummies(titanic, columns = ['sexDummies'], prefix = ['D'])
titanic.head()

### Ejercicio 2:

1. Agrupa el conjunto de datos por género (`sexo`) y calcula la edad media (`edad`).
2. Agrupar el conjunto de datos por la columna familia (`familia`) y contar el número de registros por grupo.
3. De las familias, ¿cuántas sobrevivieron y cuántas no?

In [ ]:
# Su código

Esto debería obtener para 1:


|_|sex| age | 
|-----|-----|-----|
|0|	female|	27.915709|
|1|male|30.726645|

In [ ]:
# Su código

Esto debería obtener para 2:

|_|family| count | 
|-----|-----|-----|
|0|	0|	547|
|1|1|344|

In [ ]:
# Su código...

Esto debería obtener para 3:

|_|family| survived | count |
|-----|-----|-----|-----|
|0|	0|	0| 384|
|1|	0|	1| 163|
|2|	1|	0| 165|
|3|	1|	1| 179|